File: typescript/src/core/puzzle3d.ts
import { ExactCoverCounter } from './exact-cover-counter';

type Cube = [number, number, number];

interface Placement {
  piece: number;
  cells: number[];
}

export class Puzzle3D {
  private static readonly PIECES = 12;
  private static readonly PENTOMINOES: number[][][] = [
    [[1], [1], [1], [1, 1]],
    [[1, 1], [1], [1, 1]],
    [[0, 1, 1], [1, 1], [0, 1]],
    [[0, 1], [1, 1, 1], [0, 1]],
    [[1, 1, 1, 1], [0, 0, 1]],
    [[0, 1], [1, 1], [1], [1]],
    [[0, 0, 1], [0, 1, 1], [1, 1]],
    [[1], [1, 1], [1, 1]],
    [[0, 0, 1], [1, 1, 1], [1]],
    [[0, 0, 1], [0, 0, 1], [1, 1, 1]],
    [[0, 0, 1], [1, 1, 1], [0, 0, 1]],
    [[1, 1, 1, 1, 1]],
  ];

  readonly grid: number[];
  totalSolutions = 0;
  triedPieces = 0;

  private readonly cellCount: number;
  private readonly placements: Placement[] = [];
  private readonly orientations: Cube[][][];

  constructor(
    private readonly rows: number,
    private readonly columns: number,
    private readonly depth: number,
  ) {
    if (
      !Number.isInteger(rows) ||
      !Number.isInteger(columns) ||
      !Number.isInteger(depth) ||
      rows < 1 ||
      columns < 1 ||
      depth < 1
    ) {
      throw new Error('All dimensions must be positive integers');
    }

    this.cellCount = rows * columns * depth;
    this.grid = new Array(this.cellCount).fill(0);
    this.orientations = Puzzle3D.PENTOMINOES.map((layout) =>
      Puzzle3D.buildOrientations(layout),
    );
    this.buildPlacements();
  }

  solve(): number {
    this.totalSolutions = 0;
    this.triedPieces = 0;
    this.grid.fill(0);

    if (this.cellCount !== Puzzle3D.PIECES * 5) {
      return 0;
    }

    const exactCoverRows = this.placements.map((placement) => [
      ...placement.cells,
      this.cellCount + placement.piece,
    ]);
    const result = ExactCoverCounter.findOne(
      this.cellCount + Puzzle3D.PIECES,
      exactCoverRows,
    );
    this.triedPieces = result.triedRows;
    this.totalSolutions = result.solutions;

    for (const rowIndex of result.selectedRows) {
      const placement = this.placements[rowIndex];
      for (const cell of placement.cells) {
        this.grid[cell] = placement.piece + 1;
      }
    }
    return this.totalSolutions > 0 ? 1 : 0;
  }

  private buildPlacements(): void {
    if (this.cellCount !== Puzzle3D.PIECES * 5) {
      return;
    }

    for (let piece = 0; piece < Puzzle3D.PIECES; piece++) {
      for (const orientation of this.orientations[piece]) {
        const maxRow = Math.max(...orientation.map((cube) => cube[0]));
        const maxColumn = Math.max(...orientation.map((cube) => cube[1]));
        const maxDepth = Math.max(...orientation.map((cube) => cube[2]));

        for (let rowOffset = 0; rowOffset + maxRow < this.rows; rowOffset++) {
          for (
            let columnOffset = 0;
            columnOffset + maxColumn < this.columns;
            columnOffset++
          ) {
            for (
              let depthOffset = 0;
              depthOffset + maxDepth < this.depth;
              depthOffset++
            ) {
              const cells = orientation.map(([row, column, depth]) => {
                const placedRow = rowOffset + row;
                const placedColumn = columnOffset + column;
                const placedDepth = depthOffset + depth;
                return (
                  (placedRow * this.columns + placedColumn) * this.depth +
                  placedDepth
                );
              });
              this.placements.push({ piece, cells });
            }
          }
        }
      }
    }
  }

  private static buildOrientations(layout: number[][]): Cube[][] {
    const cubes: Cube[] = [];
    for (let row = 0; row < layout.length; row++) {
      for (let column = 0; column < layout[row].length; column++) {
        if (layout[row][column] === 1) {
          cubes.push([row, column, 0]);
        }
      }
    }

    const permutations: Cube[] = [
      [0, 1, 2],
      [0, 2, 1],
      [1, 0, 2],
      [1, 2, 0],
      [2, 0, 1],
      [2, 1, 0],
    ];
    const orientations: Cube[][] = [];
    const seen = new Set<string>();

    for (const permutation of permutations) {
      const permutationSign = Puzzle3D.permutationSign(permutation);
      for (const firstSign of [-1, 1]) {
        for (const secondSign of [-1, 1]) {
          for (const thirdSign of [-1, 1]) {
            if (
              permutationSign * firstSign * secondSign * thirdSign !==
              1
            ) {
              continue;
            }

            const signs = [firstSign, secondSign, thirdSign];
            const transformed = cubes.map(
              (cube): Cube => [
                signs[0] * cube[permutation[0]],
                signs[1] * cube[permutation[1]],
                signs[2] * cube[permutation[2]],
              ],
            );
            Puzzle3D.normalizeAndSort(transformed);
            const key = transformed.map((cube) => cube.join(',')).join(';');
            if (!seen.has(key)) {
              seen.add(key);
              orientations.push(transformed);
            }
          }
        }
      }
    }
    return orientations;
  }

  private static permutationSign(permutation: Cube): number {
    let inversions = 0;
    for (let first = 0; first < permutation.length; first++) {
      for (let second = first + 1; second < permutation.length; second++) {
        if (permutation[first] > permutation[second]) {
          inversions++;
        }
      }
    }
    return inversions % 2 === 0 ? 1 : -1;
  }

  private static normalizeAndSort(cubes: Cube[]): void {
    const minimum: Cube = [Infinity, Infinity, Infinity];
    for (const cube of cubes) {
      for (let axis = 0; axis < 3; axis++) {
        minimum[axis] = Math.min(minimum[axis], cube[axis]);
      }
    }
    for (const cube of cubes) {
      for (let axis = 0; axis < 3; axis++) {
        cube[axis] -= minimum[axis];
      }
    }
    cubes.sort((left, right) => {
      for (let axis = 0; axis < 3; axis++) {
        if (left[axis] !== right[axis]) {
          return left[axis] - right[axis];
        }
      }
      return 0;
    });
  }
}

File: go/core/puzzle3d.go
package core

import (
	"context"
	"errors"
	"sort"
	"strconv"
	"strings"
)

const (
	Puzzle3DPieces = 12
	cellsPerPiece  = 5
)

type cube struct {
	row, column, depth int
}

type placement3D struct {
	piece int
	cells []int
}

type Puzzle3D struct {
	rows, columns, depth int
	cellCount            int
	placements           []placement3D

	Grid           []int
	TotalSolutions int
	TriedPieces    int
}

func NewPuzzle3D(rows, columns, depth int) (*Puzzle3D, error) {
	if rows <= 0 || columns <= 0 || depth <= 0 {
		return nil, errors.New("all dimensions must be positive integers")
	}

	puzzle := &Puzzle3D{
		rows:      rows,
		columns:   columns,
		depth:     depth,
		cellCount: rows * columns * depth,
	}
	puzzle.Grid = make([]int, puzzle.cellCount)
	puzzle.buildPlacements()
	return puzzle, nil
}

func (p *Puzzle3D) Solve(ctx context.Context, countSolutions bool) int {
	clear(p.Grid)
	p.TotalSolutions = 0
	p.TriedPieces = 0

	if p.cellCount != Puzzle3DPieces*cellsPerPiece {
		return 0
	}

	rows := make([][]int, len(p.placements))
	for i, placement := range p.placements {
		row := make([]int, 0, cellsPerPiece+1)
		row = append(row, placement.cells...)
		row = append(row, p.cellCount+placement.piece)
		rows[i] = row
	}

	limit := 1
	if countSolutions {
		limit = int(^uint(0) >> 1)
	}
	result := runExactCover(ctx, p.cellCount+Puzzle3DPieces, rows, limit)
	p.TotalSolutions = result.solutions
	p.TriedPieces = result.triedRows

	if !countSolutions {
		for _, rowIndex := range result.selectedRows {
			placement := p.placements[rowIndex]
			for _, cell := range placement.cells {
				p.Grid[cell] = placement.piece + 1
			}
		}
		if p.TotalSolutions > 0 {
			return 1
		}
	}
	return p.TotalSolutions
}

func (p *Puzzle3D) buildPlacements() {
	if p.cellCount != Puzzle3DPieces*cellsPerPiece {
		return
	}

	for piece, layout := range pentominoLayouts {
		for _, orientation := range build3DOrientations(layout) {
			maxRow, maxColumn, maxDepth := 0, 0, 0
			for _, cell := range orientation {
				maxRow = max(maxRow, cell.row)
				maxColumn = max(maxColumn, cell.column)
				maxDepth = max(maxDepth, cell.depth)
			}

			for rowOffset := 0; rowOffset+maxRow < p.rows; rowOffset++ {
				for columnOffset := 0; columnOffset+maxColumn < p.columns; columnOffset++ {
					for depthOffset := 0; depthOffset+maxDepth < p.depth; depthOffset++ {
						cells := make([]int, len(orientation))
						for i, cell := range orientation {
							row := rowOffset + cell.row
							column := columnOffset + cell.column
							depth := depthOffset + cell.depth
							cells[i] = (row*p.columns+column)*p.depth + depth
						}
						p.placements = append(p.placements, placement3D{piece: piece, cells: cells})
					}
				}
			}
		}
	}
}

func build3DOrientations(layout [][]int) [][]cube {
	var cells []cube
	for row, line := range layout {
		for column, value := range line {
			if value == 1 {
				cells = append(cells, cube{row: row, column: column})
			}
		}
	}

	permutations := [][3]int{
		{0, 1, 2}, {0, 2, 1}, {1, 0, 2},
		{1, 2, 0}, {2, 0, 1}, {2, 1, 0},
	}
	signValues := []int{-1, 1}
	seen := make(map[string]bool)
	var orientations [][]cube

	for _, permutation := range permutations {
		permutationSign := permutation3DSign(permutation)
		for _, firstSign := range signValues {
			for _, secondSign := range signValues {
				for _, thirdSign := range signValues {
					signs := [3]int{firstSign, secondSign, thirdSign}
					if permutationSign*firstSign*secondSign*thirdSign != 1 {
						continue
					}

					transformed := make([]cube, len(cells))
					for i, cell := range cells {
						values := [3]int{cell.row, cell.column, cell.depth}
						transformed[i] = cube{
							row:    signs[0] * values[permutation[0]],
							column: signs[1] * values[permutation[1]],
							depth:  signs[2] * values[permutation[2]],
						}
					}
					normalizeAndSort3D(transformed)
					key := cubesKey(transformed)
					if !seen[key] {
						seen[key] = true
						orientations = append(orientations, transformed)
					}
				}
			}
		}
	}
	return orientations
}

func permutation3DSign(permutation [3]int) int {
	inversions := 0
	for first := range permutation {
		for second := first + 1; second < len(permutation); second++ {
			if permutation[first] > permutation[second] {
				inversions++
			}
		}
	}
	if inversions%2 == 0 {
		return 1
	}
	return -1
}

func normalizeAndSort3D(cells []cube) {
	minRow, minColumn, minDepth := cells[0].row, cells[0].column, cells[0].depth
	for _, cell := range cells[1:] {
		minRow = min(minRow, cell.row)
		minColumn = min(minColumn, cell.column)
		minDepth = min(minDepth, cell.depth)
	}
	for i := range cells {
		cells[i].row -= minRow
		cells[i].column -= minColumn
		cells[i].depth -= minDepth
	}
	sort.Slice(cells, func(i, j int) bool {
		if cells[i].row != cells[j].row {
			return cells[i].row < cells[j].row
		}
		if cells[i].column != cells[j].column {
			return cells[i].column < cells[j].column
		}
		return cells[i].depth < cells[j].depth
	})
}

func cubesKey(cells []cube) string {
	var key strings.Builder
	for _, cell := range cells {
		key.WriteString(strconv.Itoa(cell.row))
		key.WriteByte(',')
		key.WriteString(strconv.Itoa(cell.column))
		key.WriteByte(',')
		key.WriteString(strconv.Itoa(cell.depth))
		key.WriteByte(';')
	}
	return key.String()
}

var pentominoLayouts = [][][]int{
	{{1}, {1}, {1}, {1, 1}},
	{{1, 1}, {1}, {1, 1}},
	{{0, 1, 1}, {1, 1}, {0, 1}},
	{{0, 1}, {1, 1, 1}, {0, 1}},
	{{1, 1, 1, 1}, {0, 0, 1}},
	{{0, 1}, {1, 1}, {1}, {1}},
	{{0, 0, 1}, {0, 1, 1}, {1, 1}},
	{{1}, {1, 1}, {1, 1}},
	{{0, 0, 1}, {1, 1, 1}, {1}},
	{{0, 0, 1}, {0, 0, 1}, {1, 1, 1}},
	{{0, 0, 1}, {1, 1, 1}, {0, 0, 1}},
	{{1, 1, 1, 1, 1}},
}
